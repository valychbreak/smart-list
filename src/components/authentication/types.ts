export type User = {
    username: string
};

export type AuthenticationData = {
    username: string,
    password: string
};
