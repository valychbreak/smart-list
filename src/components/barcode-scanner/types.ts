export interface BarcodeScanResult {
    code: string;
    format: string;
}
