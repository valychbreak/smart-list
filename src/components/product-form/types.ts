type ProductFormData = {
    barcode: string;
    barcodeType: string;
    generalName: string;

    fullName: string | null;
    country: string | null;
    companyName: string | null;
    imageUrl: string | null;
};

export default ProductFormData;
