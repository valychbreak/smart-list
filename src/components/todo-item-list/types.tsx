/* eslint-disable max-classes-per-file */
import Product from "../../entity/Product";

export class Store {
    readonly id: number;

    readonly name: string;

    constructor(id: number, name: string) {
        this.id = id;
        this.name = name;
    }

    static fromJson(json: any): Store {
        const { id, name } = json;
        return new Store(id, name);
    }
}

class TodoItem {
    id: number;

    generalName: string;

    quantity: number;

    targetProduct: Product | null;

    isBought: boolean;

    readonly productPrice: number | null;

    readonly purchasedPrice: number | null;

    private purchasedStoreInternal: Store | null;

    constructor(
        id: number,
        generalName: string,
        quantity: number,
        targetProduct: Product | null,
        isBought: boolean,
        productPrice: number | null,
        purchasedPrice: number | null,
        purchasedStore: Store | null
    ) {
        this.id = id;
        this.generalName = generalName;
        this.quantity = quantity;
        this.targetProduct = targetProduct;
        this.isBought = isBought;
        this.productPrice = productPrice;
        this.purchasedPrice = purchasedPrice;
        this.purchasedStoreInternal = purchasedStore;
    }

    static createTodoItem(
        id: number,
        generalName: string,
        productPrice: number | null = null
    ) {
        return new TodoItem(
            id,
            generalName,
            1,
            null,
            false,
            productPrice,
            null,
            null
        );
    }

    static fromProduct(product: Product, quantity?: number): TodoItem {
        const todoItem = this.createTodoItem(
            Date.now(),
            product.productGeneralName
        );

        todoItem.quantity = quantity || 1;
        todoItem.targetProduct = product;
        todoItem.isBought = false;
        return todoItem;
    }

    static fromName(name: string, quantity: number): TodoItem {
        const todoItem = this.createTodoItem(Date.now(), name);
        todoItem.quantity = quantity;
        todoItem.isBought = false;
        return todoItem;
    }

    static from(json: any): TodoItem {
        const {
            id,
            generalName,
            quantity,
            targetProduct,
            isBought,
            purchasedPrice,
            purchasedStoreInternal,
        } = json;

        let product = null;
        if (targetProduct) {
            product = Product.from(targetProduct);
        }

        let loadedStore = null;
        if (purchasedStoreInternal) {
            loadedStore = Store.fromJson(purchasedStoreInternal);
        }

        return new TodoItem(
            id,
            generalName,
            quantity,
            product,
            isBought,
            null,
            purchasedPrice,
            loadedStore
        );
    }

    setProductPrice(newPrice: number | null): TodoItem {
        const todoItem = new TodoItem(
            this.id,
            this.generalName,
            this.quantity,
            this.targetProduct,
            this.isBought,
            newPrice,
            this.purchasedPrice,
            this.purchasedStoreInternal
        );
        return todoItem;
    }

    setTargetProduct(product: Product | null): TodoItem {
        const todoItem = new TodoItem(
            this.id,
            this.generalName,
            this.quantity,
            product,
            this.isBought,
            this.productPrice,
            this.purchasedPrice,
            this.purchasedStoreInternal
        );
        return todoItem;
    }

    setPurchasedPrice(newPrice: number | null): TodoItem {
        const todoItem = new TodoItem(
            this.id,
            this.generalName,
            this.quantity,
            this.targetProduct,
            this.isBought,
            this.productPrice,
            newPrice,
            this.purchasedStoreInternal
        );
        todoItem.targetProduct = this.targetProduct;
        return todoItem;
    }

    setPurchasedStore(store: Store | null): TodoItem {
        const todoItem = this.clone();
        todoItem.purchasedStoreInternal = store;
        return todoItem;
    }

    get purchasedStore(): Store | null {
        return this.purchasedStoreInternal;
    }

    private clone() {
        return new TodoItem(
            this.id,
            this.generalName,
            this.quantity,
            this.targetProduct,
            this.isBought,
            this.productPrice,
            this.purchasedPrice,
            this.purchasedStoreInternal
        );
    }
}

export default TodoItem;
