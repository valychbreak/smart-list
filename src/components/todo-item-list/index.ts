export { default } from "./groceries-todo-page";
