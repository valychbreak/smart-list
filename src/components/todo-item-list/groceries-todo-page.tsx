import TodoItemListContextProvider from "../../pages/groceries-todo/context/TodoItemListContextProvider";
import useGroceriesTodoController, { Mode } from "./components/use-groceries-todo-controller";
import GroceriesTodoPlanningModeView from "./components/groceries-todo-planning";
import GroceriesTodoPurchasingModeView from "./components/groceries-todo-purchasing";
import GroceriesTodoToolbar from "./components/groceries-todo-tool-bar";
import { GroceriesTodoStoreContextProvider } from "./components/groceries-todo-store-context";

const GroceriesTodoPage = () => {
    const controller = useGroceriesTodoController();

    const switchMode = (isPurchaseMode: boolean) => {
        if (isPurchaseMode) {
            controller.enablePurchasingMode();
        } else {
            controller.enablePlanningMode();
        }
    };

    return (<>
        <GroceriesTodoStoreContextProvider>
            <TodoItemListContextProvider>
                <GroceriesTodoToolbar
                    currentMode={controller.currentMode}
                    onPurchaseModeToggle={switchMode} />
                {controller.currentMode === Mode.PLANNING && <>
                    <GroceriesTodoPlanningModeView />
                </>}

                {controller.currentMode === Mode.PURCHASING && <>
                    <GroceriesTodoPurchasingModeView />
                </>}
            </TodoItemListContextProvider>
        </GroceriesTodoStoreContextProvider>
    </>);
};

export default GroceriesTodoPage;
