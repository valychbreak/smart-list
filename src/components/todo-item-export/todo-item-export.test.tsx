import { shallow } from "enzyme";
import Category from "../../entity/category";
import { Store } from "../todo-item-list/types";
import ExportItem from "./export-item";
import TodoItemExport, { ExportItemsGrouped, getExportResult } from "./todo-item-export";

const groceriesCategory = { id: 0, name: "Groceries" };
const snacksCategory = { id: 1, name: "Snacks" };

const auchanStore = { id: 0, name: "Auchan" } as Store;
const zabkaStore = { id: 1, name: "Zabka" } as Store;

function createExportItem(
    generalName: string,
    purchasedPrice?: number,
    category?: Category,
    purchasedStore?: Store
) {
    return new ExportItem(
        1,
        generalName,
        1,
        true,
        null,
        purchasedPrice || null,
        purchasedStore || null,
        category || null,
        null
    );
}

describe("TodoItemExport", () => {
    const todoItemExport = () => <TodoItemExport />;

    it("should render", () => {
        const wrapper = shallow(todoItemExport());
        expect(wrapper).toMatchSnapshot();
    });

    describe("functions", () => {
        const exportItemsGrouped = (
            exportItems: ExportItem[],
            category?: Category,
            store?: Store
        ): ExportItemsGrouped => ({
            categoryName: category?.name || "undefined",
            storeName: store?.name || "undefined",
            exportItems,
        });

        it("should group by category", () => {
            const milk = createExportItem("Milk", 3.44, groceriesCategory);
            const bread = createExportItem("Bread", 2.55, groceriesCategory);
            const pringles = createExportItem("Pringles", 3.55, snacksCategory);
            const exportItems = [milk, bread, pringles];

            const result = getExportResult(exportItems);

            expect(result).toEqual([
                exportItemsGrouped([milk, bread], groceriesCategory),
                exportItemsGrouped([pringles], snacksCategory)
            ]);
        });

        it("should group by store", () => {
            const milk = createExportItem("Milk", 3.44, undefined, auchanStore);
            const bread = createExportItem("Bread", 2.55, undefined, auchanStore);
            const pringles = createExportItem("Pringles", 3.55, undefined, zabkaStore);
            const exportItems = [milk, bread, pringles];

            const result = getExportResult(exportItems);

            expect(result).toEqual([
                exportItemsGrouped([milk, bread], undefined, auchanStore),
                exportItemsGrouped([pringles], undefined, zabkaStore)
            ]);
        });

        it("should group by category and store", () => {
            const milk = createExportItem("Milk", 3.44, groceriesCategory, auchanStore);
            const bread = createExportItem("Bread", 2.55, groceriesCategory, auchanStore);
            const lays = createExportItem("Lays", 2.55, snacksCategory, auchanStore);
            const pringles = createExportItem("Pringles", 3.55, snacksCategory, zabkaStore);
            const exportItems = [milk, bread, pringles, lays];

            const result = getExportResult(exportItems);

            expect(result).toHaveLength(3);

            expect(result).toContainEqual(
                exportItemsGrouped([milk, bread], groceriesCategory, auchanStore)
            );

            expect(result).toContainEqual(
                exportItemsGrouped([lays], snacksCategory, auchanStore)
            );

            expect(result).toContainEqual(
                exportItemsGrouped([pringles], snacksCategory, zabkaStore)
            );
        });
    });
});
