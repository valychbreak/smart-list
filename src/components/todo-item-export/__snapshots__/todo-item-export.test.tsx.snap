// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItemExport should render 1`] = `
<Fragment>
  <WithStyles(ForwardRef(Dialog))
    onClose={[Function]}
    open={false}
  >
    <ExportItemEditForm
      exportItem={null}
      onClose={[Function]}
      onSubmit={[Function]}
    />
  </WithStyles(ForwardRef(Dialog))>
  <WithStyles(ForwardRef(Dialog))
    onClose={[Function]}
    open={false}
  >
    <ExportResultView
      exportItemsGrouped={Array []}
    />
  </WithStyles(ForwardRef(Dialog))>
  <WithStyles(ForwardRef(Alert))
    severity="info"
  >
    Showing only purchased items
  </WithStyles(ForwardRef(Alert))>
  <Styled(MuiBox)
    marginY={1}
  >
    <WithStyles(ForwardRef(Button))
      onClick={[Function]}
      startIcon={<Memo />}
      variant="outlined"
    >
      Show export
    </WithStyles(ForwardRef(Button))>
  </Styled(MuiBox)>
  <ExportItemList
    exportItems={Array []}
    onEdit={[Function]}
  />
</Fragment>
`;
